File: 0637) Average of Levels in Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<double> averageOfLevels(TreeNode* root) {
        vector<double> sum;
        vector<double> count;
        vector<double> ans;
        calculate(root, 0, sum, count);
        int n = sum.size();
        for(int i = 0; i < n; i++)
        {
            ans.push_back(sum[i] / count[i]);
        }
        return ans;
    }
    void calculate(TreeNode *root, int i, vector<double> &sum, vector<double> &count)
    {
        if(root == nullptr)
        {
            return;
        }
        if(i >= sum.size())
        {
            sum.push_back(0);
            count.push_back(0);
        }
        sum[i] += root->val;
        count[i]++;
        calculate(root->left, i+1, sum, count);
        calculate(root->right, i+1, sum, count);
    }
};

File: 0637) Average of Levels in Binary Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Double> averageOfLevels(TreeNode root) {
        List<Double> sum = new ArrayList<>();
        List<Double> count = new ArrayList<>();
        List<Double> ans = new ArrayList<>();
        helper(root, 0, sum, count);
        int n = sum.size();
        for(int i = 0; i < n; i++)
        {
            Double avg = sum.get(i) / count.get(i);
            ans.add(avg);
        }
        return ans;
    }
    void helper(TreeNode root, int level, List<Double> sum, List<Double> count)
    {
        if(root == null)
        {
            return;
        }
        if(level >= sum.size())
        {
            sum.add(0.0);
            count.add(0.0);
        }
        sum.set(level, sum.get(level) + root.val);
        count.set(level, count.get(level) + 1);
        helper(root.left, level+1, sum, count);
        helper(root.right, level+1, sum, count);
    }
}
